### ISSUE TITLE
Connect to a printer over USB

### ISSUE THREAD

**Issue by Indyjunk** [SoftFever/OrcaSlicer#6866]
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

I cannot connect to my printer via a USB cable.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

One of the main features Cura has over Orca slicer right now is support for printing over USB / Connecting to a printer over USB. Currently, I have my printer sitting on my desk connected to my PC. Having the ability to print over USB through Cura is simple and easy, as I don't have to download Octoprint or any other interfacing software to make any connection. Allowing users to do the same in Orca would take a key feature that Cura has over Orca away allowing users who have a 3d printer on their desk the option to simply connect with a readily available solution with no extra software or hardware solutions. 

### Describe alternatives you've considered

1. My current workaround is using Orca to generate my print path and then send the code to my printer Via Cura. This would be an unscary step if Orca supported USB printer connections.

**Comment by m2ma** [comment]
I agree, we need something like this, it can bring many users from Cura.

**Comment by flickerfly** [comment]
I'd appreciate it for my Neptune 3 Pro.

**Comment by SpiderMurphLagos** [comment]
There are thousands of users with older machines that can only use USB.
Without this feature they will have to go back to Cura.

**Comment by PavelS2180** [comment]
This feature is truly one of the most essential for those who are not just hobbyists but professionals involved in the development, assembly, and even creation of 3D printers. OrcaSlicer is already a powerful tool that supports network connectivity, but the lack of USB connection significantly limits its versatility.

Many users simply do not have the ability to modify or upgrade their printers. For them, the only available connection methods are USB or SD cards. However, in 2025, relying on SD cards as the primary method of transferring data is outdated and highly inconvenient. SD cards should be used solely for temporary storage of print files or configurations, not as a mandatory workaround for transferring files. Constantly swapping SD cards, dealing with data loss, and the inevitable wear and tear of flash storage—these are all unnecessary hassles that could be avoided if OrcaSlicer supported direct USB connectivity.

Adding USB support would make OrcaSlicer a truly complete tool, capable of replacing an entire set of fragmented solutions. It would free users from the need to rely on external software just for testing and basic printer management, making the entire workflow more seamless and efficient.

**Comment by PickyBiker** [comment]
I have 3 computers with 3D Printers in my workshop and all of them use USB for printing. Without USB support, Orca Slicer is a poor option. I know there are workarounds, but that is inconvenient and is what makes Orca Slicer a poor option. 

please fix that!

**Comment by Jwn5k** [comment]
I too would like to see Print Via USB to be implemented, as much as I love using print over network and SD cards, some very specific scenarios have risen where I have the need to print over a USB cable due to lack of functionality or resources. It bugs me that it has been in CURA forever and Orca Slicer does not have it yet.

**Comment by PavelS2180** [comment]
Hi everyone,

This feature request is clearly very important to many of us, as you can see from the number of users supporting it here. USB printing is not just a "nice-to-have" — for many users it's the only practical way to use their printers, and it would make OrcaSlicer a truly complete solution without needing to rely on Cura or extra tools.

Is this something the team is planning to work on?

If you need help testing or collecting data, many of us here (myself included) would gladly provide feedback and run tests. This feature would really improve workflow efficiency and accessibility for a lot of people.

Thanks for considering this — it would be great to hear your thoughts or any update on whether USB support might be added in the future.

**Comment by Rau771** [comment]
So it's not my fault that I cannot connect my old ender 3 directy to Orca. Well, pretty sad.
The deal is to either switch back to cura, or work through two and a half crutches and both options arent that great. Ehhhhh

**Comment by Vad0268** [comment]
I think they don't do this intentionally. Being a programmer myself I do know, that implementing serial connection is the easiest thing to do, hundred times eaisier than even a simplest tcp/ip connection... but it will never happen now. Why?

The answer is the same like the answer to why almost all new printers are coming based on heavy Klipper  machine running Linux as a host instead of light and fast Marlin?

Because they want to know what you print, where you print, and how you print. :)

All seeing eye?

Sadly enough, but that's true...

**Comment by ElementW** [comment]
@Vad0268 While implementing USB serial communication can be easy, doing so *correctly* and in a cross-platform way is significantly more complicated than a TCP/IP connection, as the former has no unifying API, but the latter has BSD sockets itself hidden by the fact [`PrintHost`](https://github.com/SoftFever/OrcaSlicer/blob/4501bf6f5727a52eca2a9bc2467a1629aaf06eb0/src/slic3r/Utils/PrintHost.hpp) and its implementations use `libcurl` since it's all HTTP.

Beyond the variety of how the serial port can be accessed (`COM` ports on Windows, `tty`/`cu` devices on Linux/macOS, direct `libusb`-based access, maybe even HIDAPI in some cases), it:
* does not identify itself uniquely for the printer (easily mixed up with any other serial interface, even non USB; [look at the bullshit Cura has to do](https://github.com/Ultimaker/Cura/blob/34a542a7b58c4e3754e6be8971af1764835df4e1/plugins/USBPrinting/USBPrinterOutputDeviceManager.py#L140) or the disastrous, still very incomplete [`Serial`](https://github.com/SoftFever/OrcaSlicer/blob/4501bf6f5727a52eca2a9bc2467a1629aaf06eb0/src/slic3r/Utils/Serial.cpp) here in Orca that's used for Flashforge printers)
* has non-trivial configuration (baudrate really) for *any* data to flow through, and some virtual USB-serial adapters do not adhere to the USB CDC ACM spec and behave badly when their config gets changed
* may require elevated privileges to access, or an install step to lower the required privs (`udev` rules on Linux and whatnot)
* has no standard encoding, framing or error correction and associated retry algorithms
* some require the GCode to be trickled in ([see Cura](https://github.com/Ultimaker/Cura/blob/34a542a7b58c4e3754e6be8971af1764835df4e1/plugins/USBPrinting/USBPrinterOutputDevice.py#L383)) as they do not have enough buffer space for a full print, meaning baby-sitting [similar to what `TCPConsole` currently does](https://github.com/SoftFever/OrcaSlicer/blob/4501bf6f5727a52eca2a9bc2467a1629aaf06eb0/src/slic3r/Utils/TCPConsole.cpp) for Flashforge and MKS

Reminder that OrcaSlicer is free software and you are not entitled to any developers' time, the availability of the USB serial feature being only a function of somebody 1. having one such printer and 2. being motivated to implement it; it has nothing to do with your conspiratorial drivel.

**Comment by Vad0268** [comment]
Here’s a polished and idiomatic version of your reply, keeping your tone
and message intact while tightening the phrasing for clarity and flow:
------------------------------

Your take is interesting — a great example of how complexity can emerge
from simple foundations 😄

Serial communication has long been a reliable and straightforward way to
connect across platforms, proving its robustness well before IPX or TCP/IP
protocols even entered the scene.

The key point is: just do it like Repetier-Host does. I’ve used it for over
a decade of printing, and its COM port-based approach has always been
solid. For those of us who grew up with it — the old-school crowd — direct
COM communication remains a convenient and intuitive way to talk to a
printer. No need to mess with TCP ports, servers, or extra setup just to
send a bit of g-code.
------------------------------

Let me know if you want to dial it up with a bit more technical authority
or keep it casual and nostalgic.

вс, 9 нояб. 2025 г. в 19:02, Céleste Wouters ***@***.***>:

**Comment by flickerfly** [comment]
Thank you @ElementW for taking the time to explain all that.

You make a great point that Serial, while a simple thing, actually just passes upstream the reality of its complexity. I hadn't considered that and now that you mention it, I'm reflecting on the time I've spent trying to figure out the correct baud rate on various serial connections, trying to solder together compatible serial connectors, manage permissions to /dev/tty and all that other stuff you mentioned. I did all that with Repetier-Host back in the day when Klipper was a glint in Kevin O'Conner's eye. How far we've come!
